### ISSUE TITLE
Python documentation does not mention that "-threads" option

### ISSUE THREAD

**Issue by tleonhardt** [swig/swig#927]
Nowhere in the SWIG Python documentation does it mention that "-threads" option for enabling multithreaded support by automatically disabling the GIL when wrapped C/C++ code is entered.  This option is absolutely critical for many common use cases, including mine.  Hence, it should be mentioned in the documentation.

If you run the following at the command line:
```bash
swig -python -help
```

You will see the following option present among the Python Options
```bash
Python Options (available with -python)
    -threads        - Add thread support for all the interface
```

I found that this "-threads" option was essential for expected multithreaded behavior when using SWIG-wrapped C++ code in a multithreaded Python application.

Oh, and thanks for making such a great tool which makes it incredibly easy to wrap C/C++ code!

**Comment by ojwb** [comment]
It is documented, but only in `CHANGES` (sadly like so much of mmatus' work).

If you fancy taking that documentation and inserting it into the manual, a PR would be great.

**Comment by tleonhardt** [comment]
@ojwb I'd be happy to create a PR with that documentation inserted in a reasonable place.  I'll try to get that in soonish.

Have the SWIG core developers considered converting the documentation to something more amenable to community and end-user contributions than hand-written HTML?  Perhaps something like [Sphinx](http://www.sphinx-doc.orgl)?

**Comment by ojwb** [comment]
I don't think moving to Sphinx has been discussed, but if we can convert the existing docs without losing formatting it sounds good to me.  But probably a topic for a fresh ticket.

**Comment by kawing-chiu** [comment]
Hi @tleonhardt @ojwb . I've read the updated Python doc, it helps a lot. But I think there is one thing that is not quite clear about threads: That is, the behavior when virtual mehod written in Python (using directors) are called from multi-thread C++ library. What should we do in this situation? Should we use `-threads`? Will swig wrapper accquire the GIL automatically before calling into the Python virtual method?

**Comment by tleonhardt** [comment]
Hi @kawing-chiu.  In the application I'm developing now, I have multi-threaded C++ code calling a function written in Python via directors in order to pass all log messages (with log level) to Python for logging via the logging module.  I actually define a base class in C++ and override that class in Python and use directors in the *.i SWIG interface file to somehow make it all magically work ;-).  

I'm using the **-threads** option in my setup.py file configuration so it applies to my entire project.  It works great and I haven't had any problems.  Before I added the **-threads** option, my code didn't work at all and I was about to switch to using Cython instead of SWIG until I found out about **-threads** via a Google search.  So, yes, I would strongly encourage using **-threads** whenever using multithreaded C/C++ code.  That being said, I honestly don't understand how directors work, so maybe someone more knowledgeable about SWIG may have something to add to this conversation.

@kawing-chiu How would you suggest we update the documentation?  I'm happy to do it, I just am not sure what to add to help clarify this situation.

**Comment by kawing-chiu** [comment]
Hi, @tleonhardt . I just spent an afternoon digging into the code swig generates. Here's what I found: if `-threads` is not used, no `PyGILState_Ensure()`/`PyGILState_Release()` or `PyEval_SaveThread()`/`PyEval_RestoreThread()` code is inserted. With `-threads`, they are inserted everywhere. For example, `PyGILState_Ensure()`/`PyGILState_Release()` is inserted in director virtual methods. That's why your code simply works after adding the option.

Also, anywhere where pure C++ code is about to call, `PyEval_SaveThread()`/`PyEval_RestoreThread()` is inserted. This last piece is not quite favorable. It is just overkill in many situations and in my opinion will likely hinder the performance (of course it also depends on your code). I think only the C++ code that is going to do I/O needs this.

So, to sum up, the final scheme I use now is to disable the latter by default:
```
%module(threads="1")
...
%feature("nothreadallow");
```
I think this is enough for my purpose (protecting the virtual director methods calls from C++). If any swig developer passes by please correct me if I'm wrong:)

As for the doc, I think it's good to mention that by default, director virtual methods are not protected by gil locks, and if locks are needed, the thread directives can be used.

**Comment by webpolis** [comment]
gosh.. I spent quite an important amount of time wondering why my app's UI was blocked once the Swig generated python extension was running.. this made it. Thanks!
